Close block windows when player is too far away

CoderJoeW opened the pull request:
## Introduction
If a player is inside of a container window and is moved over 6 blocks from the container the window will automatically be closed

In vanilla minecraft when I was testing this it closed the window after be moving 12 blocks(I could have miscounted) so that is what I implemented here

By running the following this change can be disabled

`$inventory->setMaxDistanceFromContainer(0)`

### Relevant issues
This is a resolution for the issue

#6349 

### Test

https://file.io/mphsWsJrFfcy

dependabot[bot] committed:
Bump symfony/filesystem from 6.4.7 to 6.4.9

Bumps [symfony/filesystem](https://github.com/symfony/filesystem) from 6.4.7 to 6.4.9.
- [Release notes](https://github.com/symfony/filesystem/releases)
- [Changelog](https://github.com/symfony/filesystem/blob/7.1/CHANGELOG.md)
- [Commits](https://github.com/symfony/filesystem/compare/v6.4.7...v6.4.9)

---
updated-dependencies:
- dependency-name: symfony/filesystem
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
Merge pull request #6381 from pmmp/dependabot/composer/symfony/filesystem-6.4.9

CoderJoeW committed:
If player is more then 12 blocks from container close container window

CoderJoeW committed:
fix code style issues

CoderJoeW committed:
Merge branch 'minor-next' into 6349-container-window-not-closed

jasonw4331 commented on src/player/Player.php:
```diff
@@ -1374,6 +1378,14 @@ protected function processMostRecentMovements() : void{
 			$this->sendPosition($this->location, $this->location->getYaw(), $this->location->getPitch(), MovePlayerPacket::MODE_RESET);
 		}
+
+		if ($this->openContainerPosition !== null) {
```
There should be an API for plugins to be able to disable this behavior on their custom inventories

jasonw4331 reviewed (changes requested):
The composer lock file needs removed from the PR and some other API might be nice, but it looks clean.

ShockedPlot7560 commented on src/player/Player.php:
```diff
@@ -1374,6 +1378,14 @@ protected function processMostRecentMovements() : void{
+		if ($this->openContainerPosition !== null) {
+			$distance = $this->location->distance($this->openContainerPosition);
+
+			if ($distance > self::MAX_DISTANCE_FROM_CONTAINER) {
```
Should this distance be configurable by each inventory in the case of plugin customization?

nicholass003 commented on src/player/Player.php:
```diff
@@ -2646,6 +2658,12 @@ public function setCurrentWindow(Inventory $inventory) : bool{
 		$this->currentWindow = $inventory;
+
+		if ($inventory instanceof BlockInventory) {
+			$block = $inventory->getHolder();
```
perhaps this should be named $holder for consistent code

CoderJoeW committed:
hopefully fixes issue with composer.lock file being changed

CoderJoeW committed:
fixes changes to composer.lock

CoderJoeW committed:
change variable name to holder for consistency

CoderJoeW commented:
I am not super familiar with the API implementation so that might take me a min to figure out

Do you think we should have some method like setWindowCloseDistance and setting it to 0 would disable the functionality?

CoderJoeW committed:
add max distance to inventory interface

CoderJoeW committed:
add properties and methods to override interface max distance

CoderJoeW committed:
use max distance modifier from inventory class

CoderJoeW committed:
Fix incorrect class check

CoderJoeW committed:
code style fixes

CoderJoeW commented:
@ShockedPlot7560 @nicholass003 @jasonw4331 

I believe I addressed all concerns and included a video of testing the server against a vanilla world

I found that in further testing vanilla seems to kick you out of an inventory window > 6 blocks distance so thats what I also updated this PR to do

ShockedPlot7560 commented on src/player/Player.php:
```diff
@@ -296,6 +298,9 @@ public static function isValidUserName(?string $name) : bool{
 
 	protected ?SurvivalBlockBreakHandler $blockBreakHandler = null;
 
+	protected ?SimpleInventory $openContainerInventory = null;
```
We already have this information

ShockedPlot7560 commented on src/player/Player.php:
```diff
@@ -2646,6 +2659,17 @@ public function setCurrentWindow(Inventory $inventory) : bool{
 		$inventory->onOpen($this);
 		$this->currentWindow = $inventory;
+
+		if ($inventory instanceof SimpleInventory) {
+			if ($inventory->getMaxDistanceFromContainer() > 0) {
```
I think we can introduce a new interface to give the `getMaxDistanceFromContainer` to avoid checking SimpleInventory which is not guaranteed

CoderJoeW committed:
Create new interface for proximity restrictions

CoderJoeW committed:
update code to use ProximityRestricted interface

CoderJoeW committed:
remove old code

CoderJoeW committed:
updated based on new interface

CoderJoeW committed:
remove unused code

CoderJoeW committed:
correct function name

CoderJoeW committed:
fix typo

CoderJoeW committed:
correct errors

ShockedPlot7560 commented on src/block/ProximityRestricted.php:
```diff
@@ -0,0 +1,39 @@
+	 */
+	public function getMaxDistance() : int;
+
+	/**
+	 * Returns the max distance the player can be away
+	 */
+	public function setMaxDistance(int $maxDistance) : void;
```
I don't think we need this in the API, its more a readonly value

CoderJoeW replied:
@ShockedPlot7560 Maybe so but it never hurts to give people the ability to change it

ShockedPlot7560 replied:
Any other point of view on that ?

dktapps replied:
I'd prefer to have it read-only, and then explore adding modification support later if people ask for it. Mutability always causes headaches in ways you don't expect.

ShockedPlot7560 commented on src/block/ProximityRestricted.php:
```diff
@@ -0,0 +1,39 @@
+
+declare(strict_types=1);
+
+namespace pocketmine\block;
+
+interface ProximityRestricted {
```
This need to be placed in the inventory namespace

ShockedPlot7560 replied:
And i'm not a big fan of this interfaces name

CoderJoeW replied:
@ShockedPlot7560 Any suggestions on the name?

ShockedPlot7560 commented on src/block/ProximityRestricted.php:
```diff
@@ -0,0 +1,39 @@
+
+	public const MAX_DISTANCE = 6;
+
+	/**
+	 * Returns the max distance the player can be away
+	 */
+	public function getMaxDistance() : int;
```
Max distance against ? need some documentation or name change

ShockedPlot7560 replied:
Could be typed with non-zero-int as if we want to disable them, just don't put the interface ?

ShockedPlot7560 committed:
moove ProximityRestricted into block/inventory

ShockedPlot7560 committed:
make ProximityRestricted inventory immutable

ShockedPlot7560 committed:
make code more clear

ShockedPlot7560 committed:
improve documentation

ShockedPlot7560 committed:
remove ProximityRestricted from BaseInventory

ShockedPlot7560 committed:
remove unrelated changes

ShockedPlot7560 committed:
create ProximityRestrictedTrait and implement it in the block inventories

ShockedPlot7560 committed:
change conditions

ShockedPlot7560 committed:
fix PHPstan

ShockedPlot7560 commented:
I've taking back this PR to try this to be merged.

I've made some changes to fix stale reviews but still has question about we have to include this utility.
For now, we introduce a new `ProximityRestricted` interface which need to be combined with the `BlockInventory` to be functionnal. We can't include this interface into the `BaseInventory` bc its linked to a block inventory and we can't include it into the `BlockInventory` without any BC breaks. Do we keep a new interface which need to be included everywhere we have a BlockInventory or anything else come in your minds ?

dktapps commented:
> Do we keep a new interface which need to be included everywhere we have a BlockInventory or anything else come in your minds ?

I think the whole thing is overengineered. Most likely the only thing needed is to check if the block is still within the player's reach distance. I don't think this should be the Block's responsibility at all.

dktapps commented:
Also worth asking if we really want to make an invasive change like this considering the upcoming #6533.

dktapps commented:
Closing this as I think the solution is way too complicated for the requirements.
